cpp/test/linalg/matrix_vector_op.cu

```diff
@@ -20,228 +20,229 @@
-typedef MatVecOpTest<double, int> MatVecOpTestD_i32;
-TEST_P(MatVecOpTestD_i32, Result)
-{
-  ASSERT_TRUE(devArrMatch(out_ref.data(),
-                          out.data(),
-                          params.rows * params.cols,
-                          CompareApprox<double>(params.tolerance)));
-}
-INSTANTIATE_TEST_SUITE_P(MatVecOpTests, MatVecOpTestD_i32, ::testing::ValuesIn(inputsd_i32));
+#define MV_EPS_F 0.00001f
+#define MV_EPS_D 0.0000001
 
-const std::vector<MatVecOpInputs<double, size_t>> inputsd_i64 = {
-  {0.0000001, 2500, 250, false, false, false, 1234ULL},
-  {0.0000001, 2500, 250, false, false, true, 1234ULL}};
-typedef MatVecOpTest<double, size_t> MatVecOpTestD_i64;
-TEST_P(MatVecOpTestD_i64, Result)
-{
-  ASSERT_TRUE(devArrMatch(out_ref.data(),
-                          out.data(),
-                          params.rows * params.cols,
-                          CompareApprox<double>(params.tolerance)));
-}
-INSTANTIATE_TEST_SUITE_P(MatVecOpTests, MatVecOpTestD_i64, ::testing::ValuesIn(inputsd_i64));
+/*
+ * This set of tests covers cases where all the types are the same.
+ */
+
+const std::vector<MatVecOpInputs<int>> inputs_i32 =
+  raft::util::itertools::product<MatVecOpInputs<int>>(
+    {1024}, {32, 64}, {true, false}, {true, false}, {0, 1, 2}, {0, 1, 2}, {1234ULL});
+const std::vector<MatVecOpInputs<int64_t>> inputs_i64 =
+  raft::util::itertools::product<MatVecOpInputs<int64_t>>(
+    {2500}, {250}, {false}, {false}, {0, 1}, {0, 1}, {1234ULL});
+
+template <typename T>
+struct Add1Vec {
+  static constexpr bool useTwoVectors = false;
+  HDI T operator()(T a, T b) const { return a + b; };
+};
+template <typename T>
+struct Add2Vec {
+  static constexpr bool useTwoVectors = true;
+  HDI T operator()(T a, T b, T c) const { return a + b + c; };
+};
+
+typedef MatVecOpTest<Add1Vec<float>, float, int> MatVecOpTestF_i32_add1vec;
+typedef MatVecOpTest<Add2Vec<float>, float, int> MatVecOpTestF_i32_add2vec;
+typedef MatVecOpTest<Add1Vec<float>, float, int64_t> MatVecOpTestF_i64_add1vec;
+typedef MatVecOpTest<Add2Vec<float>, float, int64_t> MatVecOpTestF_i64_add2vec;
+typedef MatVecOpTest<Add1Vec<double>, double, int> MatVecOpTestD_i32_add1vec;
+typedef MatVecOpTest<Add2Vec<double>, double, int> MatVecOpTestD_i32_add2vec;
+typedef MatVecOpTest<Add1Vec<double>, double, int64_t> MatVecOpTestD_i64_add1vec;
+typedef MatVecOpTest<Add2Vec<double>, double, int64_t> MatVecOpTestD_i64_add2vec;
+
+MVTEST(MatVecOpTestF_i32_add1vec, inputs_i32, MV_EPS_F);
+MVTEST(MatVecOpTestF_i32_add2vec, inputs_i32, MV_EPS_F);
+MVTEST(MatVecOpTestF_i64_add1vec, inputs_i64, MV_EPS_F);
+MVTEST(MatVecOpTestF_i64_add2vec, inputs_i64, MV_EPS_F);
+MVTEST(MatVecOpTestD_i32_add1vec, inputs_i32, MV_EPS_D);
+MVTEST(MatVecOpTestD_i32_add2vec, inputs_i32, MV_EPS_D);
+MVTEST(MatVecOpTestD_i64_add1vec, inputs_i64, MV_EPS_D);
+MVTEST(MatVecOpTestD_i64_add2vec, inputs_i64, MV_EPS_D);
+
+/*
+ * This set of tests covers cases with different types.
```

Comment:
It would be nice to test the case of `int8_t` matrix with float input, so that we cover a different ratio of input data sizes. Since all checks pass here, this need not hold up the current PR, and could be implemented separately.